allow defining the element used when hydrating svelte components

douglasward opened the pull request:
I found myself wanting to use a svelte component as a shortcode inside a table. This didn't work because the `div` element is used as the base element for the svelte component during the hydration preparation. 

This pull request aims to allow the setting of a custom element to be used instead and if absent the default `div` element is used.

Example shortcode usage:
```html
<table>
  {%svelteComponent name='TableRows' props='{}' options='{"element": "tbody"}' /%}
</table>
```

This worked in the project I would need it for, but before I go any further with it (including updating the tests) I wanted to check in with you to find out if this is something you would consider merging and what the prerequisites would be.

nickreese commented on src/partialHydration/mountComponentsInHtml.ts:
```diff
@@ -14,7 +14,7 @@ export default function mountComponentsInHtml({ page, html, hydrateOptions }): s
   // sometimes svelte adds a class to our inlining.
   const matches = outputHtml.matchAll(
-    /<div class="ejs-component[^]*?" data-ejs-component="([A-Za-z]+)" data-ejs-props="({[^]*?})" data-ejs-options="({[^]*?})"><\/div>/gim,
+    /<.+? class="ejs-component[^]*?" data-ejs-component="([A-Za-z]+)" data-ejs-props="({[^]*?})" data-ejs-options="({[^]*?})"><\/.+?>/gim,
```
Only thing I'm unsure of is if there is a way we can ensure that these two tags match. My regex isn't amazing, but I'm always concerned about greedy calls.

vollnhals replied:
Should be possible with a backreference. I also replaced `.+?` with `\S+` to have a narrower match.

```suggestion
    /<(\S+) class="ejs-component[^]*?" data-ejs-component="([A-Za-z]+)" data-ejs-props="({[^]*?})" data-ejs-options="({[^]*?})"><\/\1>/gim,
```

douglasward replied:
Thanks @vollnhals, I have implement this in https://github.com/Elderjs/elderjs/pull/221/commits/e74afb79c1e20d4b99cab6d181345b1927118226

nickreese reviewed (commented):
I like it. Great job, makes a ton of sense. Something I've considered and even discussed with the team.

Tests to cover the functionality and if there is a way to ensure that the regex doesn't get overly greedy that would be ideal.

nickreese commented:
@douglasward This is good to merge once we get all of the tests working. I've also written tentative docs for this functionality.

douglasward commented:
@nickreese that sounds great! Sorry for the slow reply, it was one of those weeks...

I have made a start updating the tests. Lots were failing because of the new `JSON.parse` that we do in `inlinePreprocessedSvelteComponent` to merge the passed in string options with the default - it wasn't being passed valid JSON. Not sure if this is just a problem with how the tests are set up, or whether this is going to cause problems with people's projects as well.

The `DATEPICKER` tests are still failing and I haven't worked out why yet. Here is an example of a failing expect:
```
    Expected: "<div class=\"svelte-datepicker\"><div class=\"datepicker-component\" id=\"datepickerSwrzsrVDCd\"><div>DATEPICKER</div></div></div>"
    Received: "<<div class=\"datepicker-component\" id=\"datepickerSwrzsrVDCd\"><div>DATEPICKER</div></div> class=\"svelte-datepicker\"><div class=\"ejs-component\" data-ejs-component=\"Datepicker\" data-ejs-props=\"{ \"a\": \"b\" }\" data-ejs-options=\"{ \"loading\": \"lazy\", \"element\": \"div\" }\"></div></div>"
```

Let me know if you have any ideas or pointers.

douglasward commented:
@nickreese I have found and fixed the cause of the failing DATEPICKER tests... I accidentally bumped a match index that was before the index of match I had introduced, sorry. 

So tests are all green. Have you had a chance to look over that changes I have made to them?

nickreese commented:
@douglasward I have. Whole family got sick with COVID. Still fighting it. Thanks for getting the tests passing. Will look at it now.
